proxy/shadow_test.go

```diff
@@ -0,0 +1,405 @@
+package proxy
+
+import (
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	stdlibhttptest "net/http/httptest"
+	"net/url"
+	"os"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/zalando/skipper/filters"
+	"github.com/zalando/skipper/filters/scheduler"
+	"github.com/zalando/skipper/filters/tee"
+	"github.com/zalando/skipper/metrics/metricstest"
+	"github.com/zalando/skipper/net/httptest"
+	"github.com/zalando/skipper/predicates/primitive"
+	teepred "github.com/zalando/skipper/predicates/tee"
+	"github.com/zalando/skipper/predicates/traffic"
+	"github.com/zalando/skipper/routing"
+	"github.com/zalando/skipper/routing/testdataclient"
+	sched "github.com/zalando/skipper/scheduler"
+)
+
+func TestShadowSingle(t *testing.T) {
+	s := stdlibhttptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Add("Backend", "main")
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte("OK"))
+	}))
+	defer s.Close()
+
+	shadow := stdlibhttptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Add("Backend", "shadow")
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte("OK"))
+	}))
+	defer shadow.Close()
+
+	metrics := &metricstest.MockMetrics{}
+	reg := sched.RegistryWith(sched.Options{
+		Metrics:                metrics,
+		EnableRouteFIFOMetrics: true,
+	})
+	defer reg.Close()
+
+	fr := make(filters.Registry)
+	fr.Register(tee.NewTeeLoopback())
+	fr.Register(scheduler.NewFifo())
+
+	doc := fmt.Sprintf(`
+main: PathSubtree("/")
+-> "%s";
+split: PathSubtree("/") && Traffic(0.5)
+-> teeLoopback("test")
+-> "%s";
+shadow: PathSubtree("/") && Tee("test") && True()
+-> "%s";
```

Comment:
Maybe if we indent inside of string block it would look nicer?

Reply:
done